src/ipc/handler.c

```diff
@@ -1371,158 +1306,94 @@ void ipc_cmd(struct sof_ipc_cmd_hdr *hdr)
+	if (high_priority) {
+		ret = ipc_platform_send_msg(msg);
+		if (!ret)
+			goto out;
+	}
 
-		/* iterate host message list for searching */
-		list_for_item(plist, &ipc->msg_list) {
-			msg = container_of(plist, struct ipc_msg, list);
-			if (msg->header == posn->rhdr.hdr.cmd) {
-				old_posn = (struct sof_ipc_stream_posn *)
-					   msg->tx_data;
-				if (old_posn->comp_id == posn->comp_id)
-					return msg;
-			}
-			platform_shared_commit(msg, sizeof(*msg));
-		}
-		break;
-	default:
-		break;
+	/* add to queue unless already there */
+	if (list_is_empty(&msg->list)) {
+		if (high_priority)
+			list_item_prepend(&msg->list, &ipc->msg_list);
+		else
+			list_item_append(&msg->list, &ipc->msg_list);
 	}
 
-	/* no match */
-	return NULL;
+out:
+	spin_unlock_irq(&ipc->lock, flags);
 }
 
-static inline struct ipc_msg *ipc_glb_trace_message_find(struct ipc *ipc,
-	struct sof_ipc_dma_trace_posn *posn)
+void ipc_stream_position_msg_send(struct ipc_msg *msg,
+				  const struct comp_dev *dev,
+				  struct sof_ipc_stream_posn *posn)
 {
-	struct list_item *plist;
-	struct ipc_msg *msg = NULL;
-	uint32_t cmd;
+	posn->rhdr.hdr.cmd = SOF_IPC_GLB_STREAM_MSG | SOF_IPC_STREAM_POSITION |
+		dev_comp_id(dev);
+	posn->rhdr.hdr.size = sizeof(*posn);
+	posn->comp_id = dev_comp_id(dev);
 
-	/* Check whether the command is expected */
-	cmd = iCS(posn->rhdr.hdr.cmd);
+	mailbox_stream_write(dev->pipeline->posn_offset, posn, sizeof(*posn));
 
-	switch (cmd) {
-	case SOF_IPC_TRACE_DMA_POSITION:
-		/* iterate host message list for searching */
-		list_for_item(plist, &ipc->msg_list) {
-			msg = container_of(plist, struct ipc_msg, list);
-			if (msg->header == posn->rhdr.hdr.cmd)
-				return msg;
-			platform_shared_commit(msg, sizeof(*msg));
-		}
-		break;
-	default:
-		break;
-	}
-
-	/* no match */
-	return NULL;
+	ipc_msg_queue(msg, posn->rhdr.hdr.cmd, posn, false);
 }
 
-static inline struct ipc_msg *msg_find(struct ipc *ipc, uint32_t header,
-	void *tx_data)
+void ipc_component_msg_send(struct ipc_msg *msg, const struct comp_dev *dev,
+			    struct sof_ipc_comp_event *event)
 {
-	uint32_t type;
-
-	/* use different sub function for different global message type */
-	type = iGS(header);
+	event->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG |
+		SOF_IPC_COMP_NOTIFICATION | dev_comp_id(dev);
+	event->rhdr.hdr.size = sizeof(*event);
+	event->src_comp_type = dev_comp_type(dev);
+	event->src_comp_id = dev_comp_id(dev);
 
-	switch (type) {
-	case SOF_IPC_GLB_STREAM_MSG:
-		return ipc_glb_stream_message_find(ipc,
-			(struct sof_ipc_stream_posn *)tx_data);
-	case SOF_IPC_GLB_TRACE_MSG:
-		return ipc_glb_trace_message_find(ipc,
-			(struct sof_ipc_dma_trace_posn *)tx_data);
-	default:
-		/* not found */
-		return NULL;
-	}
+	ipc_msg_queue(msg, event->rhdr.hdr.cmd, event, true);
 }
 
-void ipc_prepare_host_message(struct ipc_msg *msg, uint32_t header,
-			      void *tx_data, size_t tx_bytes)
+void ipc_xrun_msg_send(struct ipc_msg *msg, const struct comp_dev *dev,
```

Comment:
I am not sure why there is a dedicated function needed for each specific message. Looks like some generic ipc_prepare(...) called by a client could do the work?